File: 백준/Gold/1041. 주사위/주사위.cc
#include <iostream>
#include <algorithm>

using namespace std;

long long N, A, B, C, D, E, F;
int main() {
    cin >> N;

    cin >> A >> B >> C >> D >> E >> F;

    long long three_side = min({A + B + C, A + B + D, A + D + E, A + E + C, B + C + F, B + F + D, C + E + F, D + F + E});
    long long two_side = min({A + B, A + C, A + D, A + E, B + F, B + C, B + D, C + E, C + F, D + F, D + E, E + F});
    long long one_side = min({A, B, C, D, E, F});

    if(N == 1) cout << (A + B + C + D + E + F) - max({A, B, C, D, E, F});
    else {
        long long answer = 0;
        answer += 4 * three_side;
        answer += ((N - 1) * 4 + (N - 2) * 4) * two_side;
        answer += ((N - 1) * (N - 2) * 4 + (N - 2) * (N - 2)) * one_side;
        cout << answer;
    }
    
}

File: 백준/Gold/1041. 주사위/주사위.java
import java.util.*;

class Main {
    static boolean[][] adj = {
        {false, true, true, true, true, false},
        {true, false, true, true, false, true},
        {true, true, false, false, true, true},
        {true, true, false, false, true, true},
        {true, false, true, true, false, true},
        {false, true, true, true, true, false},
    };
    static long N;
    static long[] dice = new long[6];
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        N = sc.nextLong();
        for(int i = 0; i < dice.length; ++i) {
            dice[i] = sc.nextLong();
        }

        long three_side = Long.MAX_VALUE, two_side = Long.MAX_VALUE, one_side = Long.MAX_VALUE;

        long max_side = 0, sum = 0;
        for(int i = 0; i < dice.length; ++i) {
            one_side = Math.min(one_side, dice[i]);
            max_side = Math.max(max_side, dice[i]);
            sum += dice[i];
            for(int j = i + 1; j < dice.length; ++j) {
                if(!adj[i][j]) continue;
                two_side = Math.min(two_side, dice[i] + dice[j]);
                for(int k = j + 1; k < dice.length; ++k) {
                    if(adj[i][k] && adj[j][k]) three_side = Math.min(three_side, dice[i] + dice[j] + dice[k]);
                }
            }
        }

        if(N == 1) {
            System.out.print(sum - max_side);
        } else {
            long answer = 0;
            answer += three_side * 4;
            answer += ((N - 1) * 4 + (N - 2) * 4) * two_side;
            answer += ((N - 1) * (N - 2) * 4 + (N - 2) * (N - 2)) * one_side;
            System.out.print(answer);
        }
    }

    
}